src/main/java/com/google/devtools/build/lib/rules/proto/ProtoLangToolchainProvider.java

```diff
@@ -40,15 +42,46 @@
   @Nullable
   public abstract TransitiveInfoCollection runtime();
 
+  /**
+   * Returns a list of {@link ProtoSource}s that are already provided by the protobuf runtime (i.e.
+   * for which {@code <lang>_proto_library} should not generate bindings.
+   */
+  public abstract ImmutableList<ProtoSource> providedProtoSources();
+
+  @Deprecated
   public abstract NestedSet<Artifact> blacklistedProtos();
 
   @AutoCodec.Instantiator
+  public static ProtoLangToolchainProvider createForDeserialization(
+      String commandLine,
+      FilesToRunProvider pluginExecutable,
+      TransitiveInfoCollection runtime,
+      ImmutableList<ProtoSource> providedProtoSources,
+      NestedSet<Artifact> blacklistedProtos) {
+    return new AutoValue_ProtoLangToolchainProvider(
+        commandLine, pluginExecutable, runtime, providedProtoSources, blacklistedProtos);
+  }
+
+  public static ProtoLangToolchainProvider create(
+      String commandLine,
+      FilesToRunProvider pluginExecutable,
+      TransitiveInfoCollection runtime,
+      ImmutableList<ProtoSource> providedProtoSources) {
+    NestedSetBuilder<Artifact> blacklistedProtos = NestedSetBuilder.stableOrder();
+    for (ProtoSource protoSource : providedProtoSources) {
+      blacklistedProtos.add(protoSource.getOriginalSourceFile());
+    }
+    return new AutoValue_ProtoLangToolchainProvider(
+        commandLine, pluginExecutable, runtime, providedProtoSources, blacklistedProtos.build());
+  }
+
+  @Deprecated
   public static ProtoLangToolchainProvider create(
       String commandLine,
       FilesToRunProvider pluginExecutable,
       TransitiveInfoCollection runtime,
       NestedSet<Artifact> blacklistedProtos) {
     return new AutoValue_ProtoLangToolchainProvider(
-        commandLine, pluginExecutable, runtime, blacklistedProtos);
+        commandLine, pluginExecutable, runtime, ImmutableList.of(), blacklistedProtos);
   }
```

Comment:
I checked internally and the method may be safely deleted. And then I believe you can use NestedSet for providedProtoSources.

Reply:
removed